Proxy forward auth data

jai1 opened the pull request:
======

Currently the proxy extracts and forwards the client Principal to the broker. Client Principal is a modifiable string i.e can be changed or manufactured by the Proxy.

What we want to do instead is to send the clientAuthData to the broker - which is in most cases digitally signed. The broker will extract the client principal from the clientAuthData and reauthenticate the client.

In order to enforce this behavior we have introduced two new flags:-
- authenticateOriginalAuthData on the broker 
- forwardClientAuthData on the Proxy

======

@msb-at-yahoo

saandrews commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:
```diff
@@ -385,8 +465,11 @@ protected void handleConnect(CommandConnect connect) {
+                        connect.hasOriginalAuthData() ? connect.getOriginalAuthData() : null,
+                        connect.hasOriginalAuthMethod() ? connect.getOriginalAuthMethod() : null,
+                        connect.hasOriginalPrincipal() ? connect.getOriginalPrincipal() : null,
+                        sslSession);
```
What happens if sslSession is null ?

jai1 replied:
is sslSession is null AuthDataCommand.hasDataFromTls will return false

saandrews commented on pulsar-common/src/main/proto/PulsarApi.proto:
```diff
@@ -148,6 +148,13 @@ message CommandConnect {
+	
+	// Original auth role and auth Method that was passed 
+	// to the proxy. In this case the auth info above
+	// will the the auth of the proxy itself	
```
"will be the"?

Same for the remaining comments as well.

jai1 commented:
@saandrews - have addressed your comments 
@rdhabalia @merlimat - Please review this PR when you find time

jai1 commented:
retest this please

merlimat commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:
```diff
@@ -213,8 +221,19 @@ protected void handleLookup(CommandLookupTopic lookup) {
         if (lookupSemaphore.tryAcquire()) {
-            final String originalPrincipal = lookup.hasOriginalPrincipal() ? lookup.getOriginalPrincipal()
-                    : this.originalPrincipal;
+            String originalPrincipal;
```
This check can be done before acquiring the semaphore

merlimat replied:
As with similar changes, try to put common logic in a single method

merlimat commented on pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConfiguration.java:
```diff
@@ -59,7 +59,10 @@
-
+    // Forward client authData to Broker for re authorization
+    // make sure authentication is enabled for this to take effect
+    private boolean forwardAuthData = false;
```
Should be added to `proxy.conf`

merlimat replied:
Rename to something like `authorizationForwardCredentials` or similar

jai1 commented:
> As with similar changes, try to put common logic in a single method

@merlimat - Have addressed your comments

I feel the last commit (https://github.com/apache/incubator-pulsar/pull/1169/commits/f986185eb7e79bcd3fac360d159057f39eeafcc6) addresses your comment but makes the logic more difficult to understand

jai1 commented:
@merlimat @rdhabalia - can I merge this or do you want to take one last look.

rdhabalia commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:
```diff
@@ -215,10 +217,18 @@ protected void handleLookup(CommandLookupTopic lookup) {
+                lookup);
+        
+        if (authenticateOriginalAuthData && lookup.hasOriginalAuthData() && originalPrincipal == null) {
+            return;
```
this condition seems little tricky. if we fail to validate original principal then `validateOriginalPrincipal()` will send failure response to the client and then thread should not process anything further. but then with this condition there is a possibility to move forward and thread will do further processing if `authenticateOriginalAuthData && lookup.hasOriginalAuthData() =is false`.??

jai1 commented:
@rdhabalia - handled your comments - hope to get a +1 from you
